src/main/java/frc/robot/subsystems/Drivetrain.java

```diff
@@ -0,0 +1,206 @@
+                    "drivetrain",
+                    new CANSparkMax(RobotMap.CAN.DRIVETRAIN_LEFT_SPARKMAX_1,
+                            CANSparkMaxLowLevel.MotorType.kBrushless),
+                    new CANSparkMax(RobotMap.CAN.DRIVETRAIN_LEFT_SPARKMAX_2,
+                            CANSparkMaxLowLevel.MotorType.kBrushless),
+                    new CANSparkMax(RobotMap.CAN.DRIVETRAIN_RIGHT_SPARKMAX_1,
+                            CANSparkMaxLowLevel.MotorType.kBrushless),
+                    new CANSparkMax(RobotMap.CAN.DRIVETRAIN_RIGHT_SPARKMAX_2,
+                            CANSparkMaxLowLevel.MotorType.kBrushless));
+        }
+        return instance;
+    }
+
+    private Drivetrain(String namespaceName, CANSparkMax leftMaster, CANSparkMax leftSlave,
+                       CANSparkMax rightMaster, CANSparkMax rightSlave) {
+        super(
+                namespaceName, leftMaster, leftSlave, rightMaster, rightSlave);
+        this.gyro = new AHRS();
+        this.leftEncoder = leftMaster.getEncoder();
+        this.rightEncoder = rightMaster.getEncoder();
+        this.trapezoidProfileSettings = new TrapezoidProfileSettings(trapezoidVelocity, trapezoidAcceleration);
+        this.leftPIDSettings = new PIDSettings(kPLeft, kILeft, kDLeft, waitTimeLeft, toleranceLeft);
+        this.rightPIDSettings = new PIDSettings(kPRight, kIRight, kDRight, waitTimeRight, toleranceRight);
+        this.feedForwardSettings = new FeedForwardSettings(kS, kV, kA);
+        this.anglePIDSettings = new PIDSettings(kPAngle, kIAngle, kDAngle, waitTimeAngle, toleranceAngle);
+        this.odometry = new DifferentialDriveOdometry(gyro.getRotation2d(), // @todo make sure this is actually yaw
+                getLeftEncoderPosition(), getRightEncoderPosition());
+        this.kinematics = new DifferentialDriveKinematics(TRACK_WIDTH);
+        this.ramseteController = new RamseteController();
+        this.field2d = new Field2d();
+        configureDashboard();
+    }
+
+    @Override
+    public void configureLoop(PIDSettings leftPIDSettings, PIDSettings rightPIDSettings,
+                              FeedForwardSettings feedForwardSettings,
+                              TrapezoidProfileSettings trapezoidProfileSettings) {
+        super.configureLoop(leftPIDSettings, rightPIDSettings, feedForwardSettings, trapezoidProfileSettings);
+        leftEncoder.setPositionConversionFactor(DISTANCE_PER_PULSE);
+        leftEncoder.setVelocityConversionFactor(DISTANCE_PER_PULSE / SECONDS_IN_MINUTE);
+        rightEncoder.setPositionConversionFactor(DISTANCE_PER_PULSE);
+        rightEncoder.setVelocityConversionFactor(DISTANCE_PER_PULSE / SECONDS_IN_MINUTE);
+    }
+
+    @Override
+    public void periodic() {
+        super.periodic();
+        odometry.update(gyro.getRotation2d(),
+                getLeftEncoderPosition(), getRightEncoderPosition());
+        field2d.setRobotPose(getPose2d());
+    }
+
+    public void resetEncoders() {
```

Comment:
when resetting encoders or gyro, you must call `odometry.resetPosition()` with the new values